tests/lib/test_openstack_server.py

```diff
@@ -0,0 +1,320 @@
+
+        self.api = self.mocked_connection.return_value.__enter__.return_value
+
+        self.mock_server_list = [
+            {
+                "id": "serverid1",
+                "name": "server1",
+                "created_at": "2020-06-28T14:00:00Z",
+                "updated_at": "2020-06-28T14:00:00Z",
+                "status": "SHUTOFF",
+            },
+            {
+                "id": "serverid2",
+                "name": "server2",
+                "created_at": "2021-04-28T14:00:00Z",
+                "updated_at": "2021-06-28T14:00:00Z",
+                "status": "ERROR",
+            },
+            {
+                "id": "serverid3",
+                "name": "server3",
+                "created_at": "2021-06-28T14:00:00Z",
+                "updated_at": "2021-06-28T14:00:00Z",
+                "status": ["SHUTOFF", "ERROR"],
+            },
+        ]
+
+    def test_search_all_servers_no_project(self):
+        """
+        Tests calling search_all_servers with no project selected
+        """
+        # pylint:disable=too-few-public-methods,invalid-name,redefined-builtin
+        class ProjectMock:
+            def __init__(self, id):
+                self.id = id
+
+        self.identity_module.list_projects.return_value = [
+            ProjectMock("ID1"),
+            ProjectMock("ID2"),
+        ]
+
+        self.instance.search_all_servers(cloud_account="test", project_identifier="")
+
+        self.identity_module.list_projects.assert_called_once_with("test")
+        self.mocked_connection.assert_called_once_with("test")
+
+        self.api.list_servers.assert_has_calls(
+            [
+                mock.call(
+                    filters={
+                        "all_tenants": True,
+                        "project_id": "ID1",
+                        "limit": 10000,
+                    }
+                ),
+                mock.call(
+                    filters={
+                        "all_tenants": True,
+                        "project_id": "ID2",
+                        "limit": 10000,
+                    }
+                ),
+            ],
+            any_order=True,
+        )
+
+    def test_search_all_servers(self):
+        """
+        Tests calling search_all_servers
+        """
+        self.instance.search_all_servers(
+            cloud_account="test", project_identifier="project"
+        )
+
+        self.identity_module.find_mandatory_project.assert_called_once_with(
+            "test", project_identifier="project"
+        )
+        self.mocked_connection.assert_called_once_with("test")
+
+        self.api.list_servers.assert_called_once_with(
+            filters={
+                "all_tenants": True,
+                "project_id": self.identity_module.find_mandatory_project.return_value.id,
+                "limit": 10000,
+            }
+        )
+
+    @mock.patch("openstack_api.openstack_query.datetime", wraps=datetime)
```

Comment:
Why are we patching instead of doing dependency injection here?
Patching is alright for legacy / existing code, but it adds a burden to writing all tests in the future since we have to patch the dep each time rather than do something like `self.mocked_query`

Reply:
This is for datetime not openstack_query